src/libexpr/primops.cc

```diff
@@ -1648,13 +1671,20 @@ static void prim_readDir(EvalState & state, const PosIdx pos, Value * * args, Va
     auto attrs = state.buildBindings(entries.size());
 
     for (auto & ent : entries) {
-        if (ent.type == DT_UNKNOWN)
-            ent.type = getFileType(path + "/" + ent.name);
-        attrs.alloc(ent.name).mkString(
-            ent.type == DT_REG ? "regular" :
-            ent.type == DT_DIR ? "directory" :
-            ent.type == DT_LNK ? "symlink" :
-            "unknown");
+        auto & attr = attrs.alloc(ent.name);
+        if (ent.type == DT_UNKNOWN) {
+            auto epath = state.allocValue();
+            Path path2 = path + "/" + ent.name;
+            epath->mkString(path2);
+            attr.mkApp(&state.getBuiltin("readFileType"), epath);
```

Comment:
`state.getBuiltin` can be floated outside of the loop.

That's not great when the list is empty, so an early return would be nice.